File: Episode024/WebStencilsDemos/WebBrokerProject/C++/UtilsLogger.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "UtilsLogger.h"
#include <iostream>
#include <fstream>

//---------------------------------------------------------------------------
#pragma package(smart_init)

// Static member definitions
std::unique_ptr<TLogger> TLogger::FInstance;
std::mutex TLogger::FMutex;

// Global Logger instance
TLogger* Logger = TLogger::Instance();

//---------------------------------------------------------------------------
__fastcall TLogger::TLogger()
    : TObject()
{
    try
    {
        // Get log path from environment variable or use default
        String EnvLogPath = GetEnvironmentVariable("APP_LOG_PATH");
        if (!EnvLogPath.IsEmpty())
        {
            FLogFileName = EnvLogPath;
            FLogPath = System::Sysutils::ExtractFilePath(FLogFileName);
        }
        else
        {
            // Default to logs subdirectory next to the binary
            String BinaryPath = System::Ioutils::TPath::GetDirectoryName(ParamStr(0));
            FLogPath = System::Ioutils::TPath::Combine(BinaryPath, "logs");
            FLogFileName = System::Ioutils::TPath::Combine(FLogPath, "app.log");
        }
        
        EnsureLogDirectory();
        // Avoid calling Info() here to prevent re-entrancy/deadlock during Instance()
    }
    catch (Exception &E)
    {
        std::wcerr << L"TLogger initialization error: " << E.Message.c_str() << std::endl;
        // Fallback: disable file logging
        FLogFileName = "";
        FLogPath = "";
    }
}

//---------------------------------------------------------------------------
__fastcall TLogger::~TLogger()
{
    // Avoid logging during destruction to prevent exceptions on shutdown
}

//---------------------------------------------------------------------------
TLogger* TLogger::Instance()
{
    std::lock_guard<std::mutex> lock(FMutex);
    
    if (!FInstance)
    {
        // Construct instance; constructor avoids using the mutex/logging
        FInstance = std::unique_ptr<TLogger>(new TLogger());
        // Set global instance
        Logger = FInstance.get();
    }
    
    return FInstance.get();
}

//---------------------------------------------------------------------------
void TLogger::EnsureLogDirectory()
{
    if (!FLogPath.IsEmpty() && !System::Sysutils::DirectoryExists(FLogPath))
    {
        try
        {
            System::Sysutils::ForceDirectories(FLogPath);
        }
        catch (Exception &E)
        {
            std::wcerr << L"Failed to create log directory: " << E.Message.c_str() << std::endl;
            // Fallback: disable file logging instead of throwing
            FLogFileName = "";
            FLogPath = "";
        }
    }
}

//---------------------------------------------------------------------------
String TLogger::FormatLogEntry(const String& ALevel, const String& AMessage)
{
    return FormatDateTime("yyyy-mm-dd hh:nn:ss.zzz", Now()) + " [" + ALevel + "] " + AMessage;
}

//---------------------------------------------------------------------------
void TLogger::WriteToFile(const String& ALevel, const String& AMessage)
{
    std::lock_guard<std::mutex> lock(FMutex);
    
    try
    {
        String LogEntry = FormatLogEntry(ALevel, AMessage);
        
        // Also output to console for debugging
#ifdef _DEBUG
        std::wcout << LogEntry.c_str() << std::endl;
#endif
        
        // Write to file if configured
        if (!FLogFileName.IsEmpty())
        {
            std::wofstream logFile;
            logFile.open(FLogFileName.c_str(), std::ios::app | std::ios::out);
            if (logFile.is_open())
            {
                logFile << LogEntry.c_str() << std::endl;
                logFile.close();
            }
        }
    }
    catch (Exception &E)
    {
        std::wcerr << L"Logger write error: " << E.Message.c_str() << std::endl;
    }
    catch (...)
    {
        std::wcerr << L"Logger write error: Unknown exception" << std::endl;
    }
}

//---------------------------------------------------------------------------
void TLogger::Info(const String& AMessage)
{
    WriteToFile("INFO", AMessage);
}

//---------------------------------------------------------------------------
void TLogger::Error(const String& AMessage)
{
    WriteToFile("ERROR", AMessage);
}

//---------------------------------------------------------------------------
void TLogger::Warning(const String& AMessage)
{
    WriteToFile("WARN", AMessage);
}

//---------------------------------------------------------------------------
void TLogger::Debug(const String& AMessage)
{
#ifdef _DEBUG
    WriteToFile("DEBUG", AMessage);
#endif
}

File: Episode024/WebStencilsDemos/WebBrokerProject/Delphi/Utils.Logger.pas
unit Utils.Logger;

interface

uses
  System.SysUtils,
  System.Classes,
  System.IOUtils,
  System.DateUtils;

type
  TLogLevel = (llInfo, llError, llDebug, llWarning);
  
  TLogger = class
  private
    FLogFile: TextFile;
    FLogPath: string;
    FLogLevel: TLogLevel;
    FFileOpen: Boolean;
    FConsoleLogging: Boolean;
    procedure WriteLog(Level: TLogLevel; const Message: string);
    function GetLogLevelString(Level: TLogLevel): string;
    procedure EnsureLogDirectory;
    procedure EnsureLogFileOpen;
  public
    constructor Create;
    destructor Destroy; override;
    procedure Info(const Message: string);
    procedure Error(const Message: string);
    procedure Debug(const Message: string);
    procedure Warning(const Message: string);
    property LogPath: string read FLogPath write FLogPath;
    property LogLevel: TLogLevel read FLogLevel write FLogLevel;
    property ConsoleLogging: Boolean read FConsoleLogging write FConsoleLogging;
  end;

var
  Logger: TLogger;

implementation

const
  DEFAULT_LOG_PATH = 'logs';
  DEFAULT_LOG_FILE = 'app.log';
  LOG_LEVEL_STRINGS: array[TLogLevel] of string = ('INFO', 'ERROR', 'DEBUG', 'WARN');

constructor TLogger.Create;
var
  EnvLogPath: string;
begin
  inherited Create;
  FLogLevel := llInfo; // Default to Info level
  FFileOpen := False;
  FConsoleLogging := False;
  
  // Try to get log path from environment variable
  EnvLogPath := GetEnvironmentVariable('APP_LOG_PATH');
  if EnvLogPath <> '' then
    FLogPath := EnvLogPath
  else
    FLogPath := TPath.Combine(DEFAULT_LOG_PATH, DEFAULT_LOG_FILE);
    
  EnsureLogDirectory;
  
  try
    EnsureLogFileOpen;
    WriteLog(llInfo, 'Logger initialized');
  except
    on E: Exception do
      System.Writeln(Format('Failed to initialize logger: %s', [E.Message]));
  end;
end;

destructor TLogger.Destroy;
begin
  try
    if FFileOpen then
    begin
      WriteLog(llInfo, 'Logger shutting down');
      CloseFile(FLogFile);
      FFileOpen := False;
    end;
  except
    on E: Exception do
      System.Writeln(Format('Error during logger shutdown: %s', [E.Message]));
  end;
  inherited;
end;

procedure TLogger.EnsureLogDirectory;
var
  LogDir: string;
begin
  LogDir := ExtractFilePath(FLogPath);
  if (LogDir <> '') and not DirectoryExists(LogDir) then
  begin
    try
      TDirectory.CreateDirectory(LogDir);
      System.Writeln(Format('Created log directory: %s', [LogDir]));
    except
      on E: Exception do
        System.Writeln(Format('Failed to create log directory: %s', [E.Message]));
    end;
  end;
end;

procedure TLogger.EnsureLogFileOpen;
begin
  if not FFileOpen then
  begin
    AssignFile(FLogFile, FLogPath);
    if not FileExists(FLogPath) then
      Rewrite(FLogFile)
    else
      Append(FLogFile);
    FFileOpen := True;
  end;
end;

function TLogger.GetLogLevelString(Level: TLogLevel): string;
begin
  Result := LOG_LEVEL_STRINGS[Level];
end;

procedure TLogger.WriteLog(Level: TLogLevel; const Message: string);
var
  LogMessage: string;
begin
  if Level < FLogLevel then
    Exit;
    
  LogMessage := Format('[%s] [%s] %s',
    [FormatDateTime('yyyy-mm-dd hh:nn:ss.zzz', Now),
     GetLogLevelString(Level),
     Message]);
     
  // Output to console if enabled
  if FConsoleLogging then
    System.WriteLn(LogMessage);

  try
    // Ensure log file is open
    EnsureLogFileOpen;
    
    if FFileOpen then
    begin
      System.Writeln(FLogFile, LogMessage);
      Flush(FLogFile); // Ensure immediate write to disk
    end;
  except
    on E: Exception do
    begin
      System.Writeln(Format('Failed to write to log file: %s', [E.Message]));
      // Try to reopen the file on next write
      FFileOpen := False;
    end;
  end;
end;

procedure TLogger.Info(const Message: string);
begin
  WriteLog(llInfo, Message);
end;

procedure TLogger.Error(const Message: string);
begin
  WriteLog(llError, Message);
end;

procedure TLogger.Debug(const Message: string);
begin
  WriteLog(llDebug, Message);
end;

procedure TLogger.Warning(const Message: string);
begin
  WriteLog(llWarning, Message);
end;

initialization
  Logger := TLogger.Create;

finalization
  Logger.Free;

end. 